data/src/main/java/com/linkedin/data/codec/entitystream/AbstractJacksonDataEncoder.java

```diff
@@ -43,61 +37,28 @@
  *
  * @author kramgopa, xma
  */
-abstract class AbstractJacksonDataEncoder implements DataEncoder
+abstract class AbstractJacksonDataEncoder extends AbstractDataEncoder
 {
-  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractJacksonDataEncoder.class);
-
-  private static final Object MAP = new Object();
-  private static final Object LIST = new Object();
-
   protected JsonFactory _jsonFactory;
   protected JsonGenerator _generator;
-  private QueueBufferedOutputStream _out;
-  private Deque<DataComplex> _stack;
-  private Deque<Iterator<?>> _iteratorStack;
-  private Deque<Object> _typeStack;
-  private WriteHandle<? super ByteString> _writeHandle;
-  private boolean _done;
-
-  private AbstractJacksonDataEncoder(JsonFactory jsonFactory, int bufferSize)
-  {
-    _jsonFactory = jsonFactory;
-    _out = new QueueBufferedOutputStream(bufferSize);
-    _stack = new ArrayDeque<>();
-    _iteratorStack = new ArrayDeque<>();
-    _typeStack = new ArrayDeque<>();
-    _done = false;
-  }
 
   protected AbstractJacksonDataEncoder(JsonFactory jsonFactory, DataMap dataMap, int bufferSize)
   {
-    this(jsonFactory, bufferSize);
-
-    _stack.push(dataMap);
-    _typeStack.push(MAP);
+    super(dataMap, bufferSize);
+    _jsonFactory = jsonFactory;
   }
 
   protected AbstractJacksonDataEncoder(JsonFactory jsonFactory, DataList dataList, int bufferSize)
   {
-    this(jsonFactory, bufferSize);
-
-    _stack.push(dataList);
-    _typeStack.push(LIST);
+    super(dataList, bufferSize);
+    _jsonFactory = jsonFactory;
   }
 
   @Override
-  public void onInit(WriteHandle<? super ByteString> wh)
+  protected Data.TraverseCallback createTraverseCallback(OutputStream out) throws IOException
   {
-    _writeHandle = wh;
-
-    try
-    {
-      _generator = _jsonFactory.createGenerator(_out);
-    }
-    catch (IOException e)
-    {
-      _writeHandle.error(e);
-    }
+    _generator = _jsonFactory.createGenerator(out);
+    return new JacksonStreamTraverseCallback(_generator);
   }
 
   protected void writeStartObject() throws IOException
```

Comment:
I find it odd that both the traverse callback and the encoder can write using _generator. Ideally all writing to _generator should be done in the traverse callback (ie no _generator member in the Encoder class). 

These protected methods wouldn't be needed as the overriding codecs can override using the traverse callback now. 
I see these are used only by the LICOR codec. Should we clean this up? This would be breaking API change, so would require a major version bump, but the code would be cleaner.
@karthikrg what do you think?

Reply:
I overrode callback not to wait for major version. should we do major version bump just for this? I can add a todo when we plan to do major version bump this can be added with it.

Reply:
I am ok with doing this breaking change and bumping major version. Agree it is cleaner. I didn't raise this before, because we've typically avoided breaking changes to the framework.

Reply:
On second thought we don't need to introduce a major version for this. Just deprecate these methods and point to the travers callback as the way to make customizations